src/riak_dt_maxreg.erl

```
@@ -0,0 +1,162 @@
+%%
+%%   http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing,
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc
+%% A register that stores the largest integer that is assigned to it
+%% It starts as `undefined`, which is a `bottom` value, but will take
+%% on any assigned value as long as it is larger than the value it had
+%% before. Only good for storing integers.
+%% @end
+
+-module(riak_dt_maxreg).
+-behaviour(riak_dt).
+
+-export([new/0, value/1, value/2, update/3, update/4, merge/2,
+         equal/2, to_binary/1, from_binary/1, parent_clock/2,
+         stats/1, stat/2]).
+
+%% EQC API
+-ifdef(EQC).
+-include_lib("eqc/include/eqc.hrl").
+-export([gen_op/0, update_expected/3, eqc_state_value/1, init_state/0, generate/0]).
+-endif.
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-export_type([maxreg/0, maxreg_op/0]).
+
+-type maxreg() :: integer() | undefined.
+
+-type maxreg_op() :: {assign, integer()}.
+
+-type nonintegral_error() :: {error, {type, {nonintegral, term()}}}.
+
+%% @doc Create a new, empty `maxreg()'
+-spec new() -> maxreg().
+new() ->
+    undefined.
+
+%% @doc The single value of a `maxreg()'.
+-spec value(maxreg()) -> term().
+value(Value) ->
+    Value.
+
+%% @doc query this `maxreg()'.
+-spec value(term(), maxreg()) -> term().
+value(_, V) ->
+    value(V).
+
+%% @doc Assign a `Value' to the `maxreg()'
+-spec update(maxreg_op(), riak_dt:actor(), maxreg()) ->
+                    {ok, maxreg()} | nonintegral_error().
+update({assign, Value}, _Actor, OldVal) when is_integer(Value) ->
+    {ok, merge(OldVal, Value)};
+update({assign, Value}, _Actor, _OldVal) ->
```

You could let the caller crash, yes, or use `error(badarg)`.